common/folderDeletionManager.go

```diff
@@ -0,0 +1,226 @@
+// Copyright Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"context"
+	"net/url"
+	"strings"
+	"sync"
+)
+
+// folderDeletionFunc should delete the folder IF IT IS EMPTY, and return true.
+// If it is not empty, false must be returned.
+// FolderDeletionManager is allowed to call this on a folder that is not yet empty.
+// In that case, FolderDeletionManager may call it again later.
+// Errors are not returned because of the delay to when deletion might happen, so
+// it's up to the func to do its own logging
+type FolderDeletionFunc func(context.Context, ILogger) bool
+
+// FolderDeletionManager handles the fact that (in most locations) we can't delete folders that
+// still contain files.  So it allows us to request deletion of a folder, and have that be attempted
+// after the last file is removed.  Note that maybe the apparent last file isn't the last (e.g.
+// there are other files, still to be deleted, in future job parts), in which case any failed deletion
+// will be retried if there's a new "candidate last child" removed.
+// Takes URLs rather than strings because that ensures correct (un)escaping, and makes it clear that we
+// don't support Windows & MacOS local paths (which have cases insensitivity that we don't support here).
+type FolderDeletionManager interface {
+
+	// RecordChildExists takes a child name and records it counts it against the child's immediate parent
```

Comment:
"records it counts it" sounds like a typo

Reply:
Yeah, must be typo because it sure doesn't sound like English! ;-) i'll fix it.